var/spack/repos/builtin/packages/zlib/package.py

```diff
@@ -25,27 +25,42 @@ class Zlib(Package):
             description='Produce position-independent code (for shared libs)')
     variant('shared', default=True,
             description='Enables the build of shared libraries.')
+    variant('debug',   default=False,
+            description='Build debug code (-g -O0)')
     variant('optimize', default=True,
             description='Enable -O2 for a more optimized lib')
 
+    conflicts('+debug+optimize')
+    conflicts('+shared', when='@:1.2.5')
+
     patch('w_patch.patch', when="@1.2.11%cce")
 
+    def patch(self):
+        if self.spec.satisfies('@:1.2.5'):
+            filter_file(r'stdio.h', 'stdio.h>\n#include <errno.h', 'gzio.c')
+
     @property
     def libs(self):
         shared = '+shared' in self.spec
         return find_libraries(
             ['libz'], root=self.prefix, recursive=True, shared=shared
         )
 
-    def setup_build_environment(self, env):
-        if '+pic' in self.spec:
-            env.append_flags('CFLAGS', self.compiler.cc_pic_flag)
-        if '+optimize' in self.spec:
-            env.append_flags('CFLAGS', '-O2')
+    def flag_handler(self, name, flags):
+        if name in ('cflags'):
```

Comment:
These are redundant parentheses, not a tuple, and it currently works only because `in` is a superset of `==`. I personally prefer to use a "list", since I know that is always going to be iterable.
```suggestion
        if name in ['cflags']:
```

However, it's not clear to me why we're using `in` instead of `==` here. Is that an artifact from a prior version of this change which chose among more than just `'cflags'`?

Finally, it is often much more clear to read methods which do an `if` statement at the top and an early `return`, allowing the rest of the method to be written at the lowest indentation level. To wit:

```suggestion
        if name != 'cflags':
            return (None, flags, None)
        if '+pic' in self.spec:
            ...
```

The `flag_handler()` API also seems strange and error-prone to me, but I'll figure that out on my own in a separate issue.